### DISCUSSION TITLE
Relax extension method and operator restrictions

### DISCUSSION THREAD

**Discussion by Reinms** [dotnet/csharplang#5478] [Language Ideas]
# Changes

There are a few parts to this idea. The first would be to allow a generic type `T<>` to declare operators for a constructed version of itself `T<U>` 
```cs
struct Container<T>
{
    public static Container<string> operator +(Container<string> lhs, Container<string> rhs) => default;
    public static Container<int> operator --(Container<int> self) => default;
}
```
These operators would of course only be applicable to instantiations of the specific types they use.

The second part would be to slightly expand the areas searched for extension methods to include the type being extended and allow the definition of extension methods for that type within itself.
```cs
class Foo
{
    public static void ExtensionMethod(this Foo self)  { }
}

class Bar<T>
{
    public static void ExtensionMethod(this Bar<T> self) { }
}
```
In addition to directly extending the type the method is defined in, it would also be allowed to extend constructed versions of itself.
```cs
class Bar<T>
{
    public static void ExtensionMethodForInts(this Bar<int> self) { }
    
    public static void ExtensionMethodFor<TInternal>(this Bar<TInternal> self}
        where TInternal : IDivisionOperators<TInternal, TInternal, TInternal>
        { }
}
```
Finally, this would also work for static abstract interface members.
```cs
interface IExample<TSelf>
    where TSelf : IExample<TSelf>
{
    static abstract void Extension(this TSelf self);
}
struct ExampleStruct : IExample<ExampleStruct>
{
    public static void Extension(this ExampleStruct self) { }
    //OR
    static void IExample<ExampleStruct>.Extension(this ExampleStruct self) { }
}
```

# Motivation

### For allowing extension methods to be defined within the type they extend
The appeal of extension methods is the ability to invoke them using either instance member syntax or static member syntax. The drawback is that the extension method is not able to be placed within the type it applies to. Currently, the two options for a hypothetical `Double` type with an extension method `AbsoluteValue()` are `myDouble.AbsoluteValue()` and `DoubleExtensions.AbsoluteValue(myDouble)`. Allowing the extension method to be defined within the type itself instead of externally only would allow `Double.AbsoluteValue(myDouble)` without needing to define the method twice.

This also sidesteps the problems with allowing extension methods everywhere and only adds one (slightly more if you also check base types) places the compiler needs to search for extension methods rather than suddenly adding every single type to the list.

### For also applying this to static abstract interface members
First, it seems like the natural thing to do if already allowing an extension to be defined within the type. More importantly though, it dramatically improves the ability to maintain type inference with extension methods and interfaces. Take the following example:

```cs
//Without allowing this in interfaces
public interface ISequence<TSelf, T>
    where TSelf : ISequence<TSelf, T>
{
    internal T[] GetValuesInternal();
}
public struct Impl<T> : ISequence<Impl<T>, T>
{
    T[] ISequence<Impl<T>, T>.GetValuesInternal() => default;
}
public static class Extensions
{
    public static T[] GetValues<TSequence, T>(this TSequence sequence)
        where TSequence : ISequence<TSequence, T>
        => sequence.GetValuesInternal();
}
public void Usage(Impl<int> instance)
{
    int[] values = instance.GetValues(); //Error, the type inference fails because the function only infers TInstance to be Impl<int> but cannot infer T to be int.
    int[] values = instance.GetValues<Impl<int>, int>(); //The 'correct' way to do it.
}



//With allowing this in interfaces
public interface ISequence<TSelf, T>
    where TSelf : ISequence<TSelf, T>
{
    internal T[] GetValuesInternal();
    
    static abstract T[] GetValues(this TSelf self);
}
public struct Impl<T> : ISequence<Impl<T>, T>
{
    T[] ISequence<Impl<T>, T>.GetValuesInternal() => default;
    
    public static T[] GetValues(this Impl<T> self) => self.GetValuesInternal();
}
public void Usage(Impl<int> instance)
{
    int[] values = instance.GetValues(); //No error now.
}
```
While this is a simple case and there are alternatives, that does not hold in more complicated cases.

### For operators on constructed versions of the type

If I have a type `ThreeOf<T>` and I want to define a `+` operator I can only do so if I constrain `T` to having a `+` operator as well. If my `ThreeOf<T>` type has some other desirable functionality then it must either be reimplemented without the `+` operator, or whoever is using it must add a `+` operator to their type even if that makes no semantic sense (Most likely throwing an exception if it is ever called)

## Rough spots

- With operator specialization it is very likely to cause people to want ways to perform that specialisation in a generic way, which would require generic operators.
- People would also very likely want to be able to perform specialisation from within an interface that has a static abstract extension method. That is flat out impossible to do without HKTs (and I think it's unrealistic to expect those to be coming any time before we're well into the triple digit C# version numbers.)

## Less rough spots

- This doesn't seem like it would be a breaking change?
- Implementation seems very straightforward. I'm not too familiar with roslyn internals but it looks like all of this would only require fairly minor adjustments to 4-5 functions.

**Comment by HaloFour** [comment]
What are the motivating use cases for either of these proposals?  Why are they combined into a single discussion instead of separately?

> The drawback is that the extension method is not able to be placed within the type it applies to.

This smells like something very different from extension methods, which are very intentionally expected to be defined outside of the type being extended due to the intent of "extending" the functionality of that type.  If you could add members directly to the type you wouldn't need extension methods.

**Reply by Reinms** [reply]
The main motivating use case for me right now is working with generic math (this is also what sorta ties the two parts together). When implementing, say, a new floating point type `Quadruple` it becomes desirable to be able to provide both a `Quadruple.Sin(myValue)` and a `myValue.Sin()` method through a single interface. When working with a `List<Quadruple>` having that method defined as an extension would allow `list.Select(Quadruple.Sin)` instead of `list.Select(q => q.Sin())` without needing to define every function in two different places.

This may not really be what extension methods were intended to do, but it does match with the way extension methods are used in a lot of codebases (IE: defining a type and then defining extension methods for that same type even though the extension methods are in the same library as the type. The extensions could have just been made part of the type but the pattern still seems to come up a lot)
